Make sender configurable when generating channel email address

prakhar1144 opened the pull request:
* Disable "Generate email address" button with a tooltip when neither the current user nor any of the bots they control has the post message permission.
* Updated UI to configure "Sender"


Fixes #31566

---

### When neither the user nor any of their bots has the permission.

<img width="417" alt="Screenshot 2025-01-02 at 7 38 21 PM" src="https://github.com/user-attachments/assets/18e73610-a1c5-45a2-8414-b8e84ef49d66" />

### A new selector at the top of this modal to choose the sender
<img width="418" alt="Screenshot 2025-02-10 at 4 15 11 PM" src="https://github.com/user-attachments/assets/bc3ccc2f-b8f1-4649-b982-65943be92d41" />

### Updated UX

* When the modal is opened, we show the email address for 'Email gateway' bot (default).
* The copy icon is now present alongside the address (see above image), it's a common UX so shouldn't be confusing ?
* When another sender is selected, we clear the address, disable the checkboxes, and focus on the "Generate email address" button to convey what should be done. See image below:
<img width="413" alt="Screenshot 2025-02-10 at 4 16 22 PM" src="https://github.com/user-attachments/assets/1c1a476e-f4f1-4240-979d-2acbdfb428d3" />



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

prakhar1144 commented:
@alya This is ready to review.

alya commented:
I think the Email Gateway option should be "Email Gateway bot" for clarity.

alya commented:
"Who will send the message?" doesn't seem clear now that I look at it in context. Let's try:

> Who should be the sender of the Zulip message sent to this channel?

And the next question becomes:

> Which parts of the email should be included in the message?

alya commented:
Let's get rid of the filter box -- it's better if we don't have it when there are just two options, which is the case that the great majority of users are going to see. I think having enough bots that you actually need it would be very rare, but if we want to make it conditional on the number of options, that would be OK.

![Screenshot 2025-02-10 at 09 21 46@2x](https://github.com/user-attachments/assets/80d3ffae-78c0-4eb9-b49e-b0070eedf0f8)

prakhar1144 commented:
Addressed the review comments:

https://github.com/zulip/zulip/pull/32382#issuecomment-2648720156,  https://github.com/zulip/zulip/pull/32382#issuecomment-2648730379, and https://github.com/zulip/zulip/pull/32382#issuecomment-2648740564:

<img width="261" alt="Screenshot 2025-02-11 at 2 00 29 AM" src="https://github.com/user-attachments/assets/beab63c8-5666-4689-89d6-1cbff0711151" />

https://github.com/zulip/zulip/pull/32382#issuecomment-2648729139:

<img width="405" alt="Screenshot 2025-02-11 at 2 00 45 AM" src="https://github.com/user-attachments/assets/42a790d6-681a-49e3-a210-3e3372e13b45" />

https://github.com/zulip/zulip/pull/32382#issuecomment-2648712062:

<img width="401" alt="Screenshot 2025-02-11 at 2 04 01 AM" src="https://github.com/user-attachments/assets/cc193c37-43e8-49a9-ad5c-c6ea9f93b810" />

---
> I think the Email Gateway option should be "Email Gateway bot" for clarity.

Used "Email Gateway Bot". "Bot" instead of "bot" as we use "Welcome Bot" etc in the name at other places.

prakhar1144 commented on web/src/stream_edit.ts:
```diff
@@ -397,45 +403,114 @@ function show_stream_email_address_modal(address: string): void {
+            sender_dropdown_widget.render();
+            $(sender_dropdown_widget.widget_selector).trigger("input");
+            dropdown.hide();
+            $("#copy_email_address_modal .tag-checkbox").off("change");
```
(comment for Tim)

I'm disabling the "change" event handler here because if not done, it leads to address being updated behind the scene (visibility: hidden). When enabling all checkboxes it leads to address wrapping on multiple line and the div containing the email address increases in size with no text displayed -- which is kinda weird UX.

Setting `display: None` will also lead to shrinking of the container div of the email-address.

I decided to disable the checkboxes first but Alya pointed out that to be not a good UX, so settled with this approach.

timabbott replied:
OK. We should just put a comment explaining this quirk.

alya commented:
Maybe this one should still be "bot" because it's not part of the name, and won't appear in the sender for those messages? See CZO discussion.

prakhar1144 commented:
Made the small tweak https://github.com/zulip/zulip/pull/32382#issuecomment-2649224389 :

<img width="245" alt="Screenshot 2025-02-11 at 2 38 07 AM" src="https://github.com/user-attachments/assets/6241646f-c2fa-44db-a94d-50375c29f13a" />

timabbott commented on web/templates/stream_settings/copy_email_address_modal.hbs:
```diff
@@ -1,6 +1,9 @@
+      label=(t 'Who should be the sender of the Zulip message sent to this channel?')}}
     <p class="question-which-parts">
-        {{t "Which parts of the email should be included in the Zulip message sent to this channel?"}}
+        {{t "Which parts of the email should be included in the message?"}}
```
I think we want to make "message" plural here, since it's a reusable email address.

timabbott commented:
I made the following changes to address the above:

``` diff
diff --git a/web/src/stream_edit.ts b/web/src/stream_edit.ts
index de932c226f..0b1b46f39d 100644
--- a/web/src/stream_edit.ts
+++ b/web/src/stream_edit.ts
@@ -413,6 +413,9 @@ function show_stream_email_address_modal(address: string, sub: StreamSubscriptio
     }
 
     function enable_tag_checkbox_change_handler(): void {
+        // Since channel email addresses encode a sender, we can only
+        // continuously update the email address when we have a current
+        // sender.
         $("#copy_email_address_modal .tag-checkbox").on("change", () => {
             const flags = get_checked_tags();
             address = get_stream_email_address(flags, address);
@@ -448,9 +451,14 @@ function show_stream_email_address_modal(address: string, sub: StreamSubscriptio
             sender_dropdown_widget.render();
             $(sender_dropdown_widget.widget_selector).trigger("input");
             dropdown.hide();
+
+            // Since channel email addresses encode a sender, we can
+            // only continuously update the email address when we have
+            // a current sender.
             $("#copy_email_address_modal .tag-checkbox").off("change");
             $(".stream-email").children().css("visibility", "hidden");
             $("#copy_email_address_modal .dialog_submit_button").trigger("focus");
+
             event.stopPropagation();
             event.preventDefault();
         }
diff --git a/web/templates/stream_settings/copy_email_address_modal.hbs b/web/templates/stream_settings/copy_email_address_modal.hbs
index d39a459f27..5c7131e07d 100644
--- a/web/templates/stream_settings/copy_email_address_modal.hbs
+++ b/web/templates/stream_settings/copy_email_address_modal.hbs
@@ -1,9 +1,9 @@
 <div class="copy-email-modal">
     {{> ../dropdown_widget_with_label
       widget_name="sender_channel_email_address"
-      label=(t 'Who should be the sender of the Zulip message sent to this channel?')}}
+      label=(t 'Who should be the sender of the Zulip messages for this email address?')}}
     <p class="question-which-parts">
-        {{t "Which parts of the email should be included in the message?"}}
+        {{t "Which parts of the emails should be included in the Zulip messages?"}}
     </p>
     {{#each tags}}
         {{#if (eq this.name "prefer-html") }}
```

timabbott commented:
The labels might benefit from further fiddling, but I think this is at least accurate.
